internal/cache/file/downloader/job.go

```diff
@@ -124,6 +126,47 @@ func (job *Job) Cancel() {
 	defer job.mu.Unlock()
 }
 
+// addSubscriber adds subscriber for download job and returns channel which is
+// notified when the download is completed at least till the subscribed offset
+// or in case of failure.
+//
+// Not concurrency safe and requires LOCK(job.mu)
+func (job *Job) addSubscriber(subscribedOffset int64) (notificationC <-chan JobStatus) {
+	subscriberC := make(chan JobStatus, 1)
+	job.subscribers.PushBack(jobSubscriber{subscriberC, subscribedOffset})
+	return subscriberC
+}
+
+// notifySubscribers notifies all the subscribers of download job in case of
+// error/cancellation or when download is completed till the subscribed offset.
+//
+// Not concurrency safe and requires LOCK(job.mu)
+func (job *Job) notifySubscribers() {
+	subItr := job.subscribers.Front()
+	for subItr != nil {
+		subItrValue := subItr.Value.(jobSubscriber)
+		nextSubItr := subItr.Next()
+		if job.status.Name == FAILED || job.status.Name == CANCELLED || job.status.Offset >= subItrValue.subscribedOffset {
```

Comment:
Just thinking out loud, the failed status here we will get after trying all the retry configured.

Reply:
Can we simplify the looping over list?

Something like this,

``` go

for e := l.Front(); e != nil; e = e.Next() {
        v := e.Value
}
```

Reply:
That will not work because we are removing elements while looping (if notified).

Reply:
Still we can do this, the only thing we need to maintain a variable as we are doing above nextSubItr.

something like this,
``` go
var next *Element
for e := l.Front(); e != nil; e = next {
    next = e.Next()
    l.Remove(e)
}
```